mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/implementation/PersistentGame.kt

```diff
@@ -14,64 +14,109 @@ import ch.epfl.sdp.mobile.application.chess.rules.Role
+              previous = this to action,
               nextPlayer = nextPlayer.other(),
-              board = nextBoard,
+              boards = boards.add(nextBoard),
           )
         }
-
-  override fun actions(position: Position): Sequence<Action> {
-    return normalizedMoves(position).map { it.first }.map {
-      Action(nextPlayer.normalize(it.from), nextPlayer.normalize(it.delta))
+      }
     }
-  }
+
+  override fun actions(position: Position): Sequence<Action> =
+      actionsAndEffects(position).map { it.first }
 
   /**
-   * Returns all the [Moves] which are available for the given position.
-   *
-   * @param position the [Position] for which the moves are queried.
-   * @return the [Moves] for the position.
+   * Returns a [Sequence] of [Action] and [Effect] that can be performed at the given [Position] by
+   * the player [nextPlayer]. Actions which would result in a check position are ignored.
    */
-  private fun normalizedMoves(position: Position): Moves {
-    val normalizedBoard = PersistentNormalizedGameBoardWithHistory(nextPlayer)
-    val normalizedFrom = nextPlayer.normalize(position)
-    val piece = normalizedBoard[normalizedFrom] ?: return emptySequence()
-    if (piece.color == Role.Adversary) return emptySequence()
-    return piece.rank.moves(normalizedBoard, normalizedFrom)
+  private fun actionsAndEffects(
+      position: Position,
+  ): Sequence<Pair<Action, Effect<Piece<Color>>>> =
+      boardSequence.allMovesAtPosition(nextPlayer, position).filter { (_, effect) ->
+        val next = effect.perform(board)
+        !(sequenceOf(next) + boardSequence).inCheck(nextPlayer)
+      }
+}
+
+/** Returns true iff the [Board] has any move available for any piece of the given [Color]. */
+private fun BoardSequence<Piece<Color>>.hasAnyMoveAvailable(color: Color): Boolean {
+  return allMoves(color).any { (_, effect) ->
+    !(sequenceOf(effect.perform(first())) + this).inCheck(color)
+  }
+}
+
+/** Returns true iff the [Board] has at least one piece of the given [Rank] and [Color]. */
+private fun <Color> Board<Piece<Color>>.hasAnyOfRankAndColor(rank: Rank, player: Color): Boolean {
+  return asSequence().any { (_, piece) -> piece.rank == rank && piece.color == player }
+}
+
+/**
+ * Returns true if and only if the player with the given color is in check.
+ *
+ * @param player the [Color] of the player for whom the check is checked.
+ */
+private fun BoardSequence<Piece<Color>>.inCheck(player: Color): Boolean {
+  return allMoves(player.other()).any { (_, effect) ->
+    !effect.perform(first()).hasAnyOfRankAndColor(King, player)
   }
 }
+
+/**
+ * Computes all the [Moves] which can be performed by a player of the given [color].
+ *
+ * @param color the [Color] of the player for whom all the moves are computed.
+ */
+private fun BoardSequence<Piece<Color>>.allMoves(color: Color): Moves<Piece<Color>> {
+  return first().asSequence().flatMap { (pos, _) -> allMovesAtPosition(color, pos) }
```

Comment:
[Question] Why are you using `first()` and not `last()` for `BoardSequence` isn't the latter supposed to be the most updated one ?

Reply:
The `Sequence` provides access to the most recent board state first. This way, since `Sequence` is a lazy collection, moves that only need to access the most recent states can iterate on the first values of the collection.